File: src/fable-library-py/fable/option.py
class Some:
    def __init__(self, value):
        self.value = value

    def __eq__(self, other):
        if self is other:
            return True

        if other is None:
            return False

        if self.value == other.value:
            return True

        return False

    def __str__(self):
        return f"Some {self.value}"

    def __repr__(self):
        return str(self)


def default_arg(opt, default_value):
    return value(opt) if opt is not None else default_value


def default_arg_with(opt, def_thunk):
    return value(opt) if opt is not None else def_thunk()


def filter(predicate, opt):
    if opt is not None:
        return opt if predicate(value(opt)) else None
    return opt


def map(mapping, opt):
    return some(mapping(value(opt))) if opt is not None else None


def map2(mapping, opt1, opt2):
    return mapping(value(opt1), value(opt2)) if (opt1 is not None and opt2 is not None) else None


def map3(mapping, opt1, opt2, opt3):
    return (
        mapping(value(opt1), value(opt2), value(opt3))
        if (opt1 is not None and opt2 is not None and opt3 is not None)
        else None
    )


def some(x):
    return Some(x) if x is None or isinstance(x, Some) else x


def value(x):
    if x is None:
        raise Exception("Option has no value")
    else:
        return x.value if isinstance(x, Some) else x


def of_nullable(x):
    return x


def to_nullable(x):
    return None if x is None else value(x)


def flatten(x):
    return None if x is None else value(x)


def to_array(opt):
    return [] if opt is None else [value(opt)]


def bind(binder, opt):
    return binder(value(opt)) if opt is not None else None


__all__ = [
    "bind",
    "defaultArg",
    "defaultArgWith",
    "flatten",
    "map",
    "map2",
    "map3",
    "ofNullable",
    "some",
    "Some",
    "toArray",
    "toNullable",
    "value",
]

File: src/fable-compiler-js/fable-library/Option.js
import { structuralHash, equals, compare } from "./Util.js";
// Using a class here for better compatibility with TS files importing Some
export class Some {
    constructor(value) {
        this.value = value;
    }
    toJSON() {
        return this.value;
    }
    // Don't add "Some" for consistency with erased options
    toString() {
        return String(this.value);
    }
    GetHashCode() {
        return structuralHash(this.value);
    }
    Equals(other) {
        if (other == null) {
            return false;
        }
        else {
            return equals(this.value, other instanceof Some ? other.value : other);
        }
    }
    CompareTo(other) {
        if (other == null) {
            return 1;
        }
        else {
            return compare(this.value, other instanceof Some ? other.value : other);
        }
    }
}
export function value(x) {
    if (x == null) {
        throw new Error("Option has no value");
    }
    else {
        return x instanceof Some ? x.value : x;
    }
}
export function unwrap(opt) {
    return opt instanceof Some ? opt.value : opt;
}
export function some(x) {
    return x == null || x instanceof Some ? new Some(x) : x;
}
export function ofNullable(x) {
    // This will fail with unit probably, an alternative would be:
    // return x === null ? undefined : (x === undefined ? new Some(x) : x);
    return x == null ? undefined : x;
}
export function toNullable(x) {
    return x == null ? null : value(x);
}
export function flatten(x) {
    return x == null ? undefined : value(x);
}
export function toArray(opt) {
    return (opt == null) ? [] : [value(opt)];
}
export function defaultArg(opt, defaultValue) {
    return (opt != null) ? value(opt) : defaultValue;
}
export function defaultArgWith(opt, defThunk) {
    return (opt != null) ? value(opt) : defThunk();
}
export function orElse(opt, ifNone) {
    return opt == null ? ifNone : opt;
}
export function orElseWith(opt, ifNoneThunk) {
    return opt == null ? ifNoneThunk() : opt;
}
export function filter(predicate, opt) {
    return (opt != null) ? (predicate(value(opt)) ? opt : undefined) : opt;
}
export function map(mapping, opt) {
    return (opt != null) ? some(mapping(value(opt))) : undefined;
}
export function map2(mapping, opt1, opt2) {
    return (opt1 != null && opt2 != null) ? mapping(value(opt1), value(opt2)) : undefined;
}
export function map3(mapping, opt1, opt2, opt3) {
    return (opt1 != null && opt2 != null && opt3 != null) ? mapping(value(opt1), value(opt2), value(opt3)) : undefined;
}
export function bind(binder, opt) {
    return opt != null ? binder(value(opt)) : undefined;
}
export function tryOp(op, arg) {
    try {
        return some(op(arg));
    }
    catch {
        return undefined;
    }
}
